core/trino-main/src/test/java/io/trino/operator/aggregation/InterpretedAggregationMaskBuilder.java

```diff
@@ -0,0 +1,138 @@
+
+    public InterpretedAggregationMaskBuilder(int... nonNullArguments)
+    {
+        this.nonNullArguments = Arrays.stream(nonNullArguments)
+                .mapToObj(NonNullArgument::new)
+                .collect(toImmutableList());
+    }
+
+    @Override
+    public AggregationMask buildAggregationMask(Page arguments, Optional<Block> optionalMaskBlock)
+    {
+        int positionCount = arguments.getPositionCount();
+
+        // if page is empty, we are done
+        if (positionCount == 0) {
+            return AggregationMask.createSelectNone(positionCount);
+        }
+
+        Block maskBlock = optionalMaskBlock.orElse(null);
+        boolean hasMaskBlock = maskBlock != null;
+        boolean maskBlockMayHaveNull = hasMaskBlock && maskBlock.mayHaveNull();
+        if (maskBlock instanceof RunLengthEncodedBlock rle) {
+            if (!testMaskBlock(rle.getValue(), maskBlockMayHaveNull, 0)) {
+                return AggregationMask.createSelectNone(positionCount);
+            }
+            // mask block is always true, so do not evaluate mask block
+            hasMaskBlock = false;
+            maskBlockMayHaveNull = false;
+        }
+
+        for (NonNullArgument nonNullArgument : nonNullArguments) {
+            nonNullArgument.reset(arguments);
+            if (nonNullArgument.isAlwaysNull()) {
+                return AggregationMask.createSelectNone(positionCount);
+            }
+        }
+
+        // if there is no mask block, and all non-null arguments do not have nulls, we are done
+        if (!hasMaskBlock && nonNullArguments.stream().noneMatch(NonNullArgument::mayHaveNull)) {
+            return AggregationMask.createSelectAll(positionCount);
+        }
+
+        // grow the selection array if necessary
+        int[] selectedPositions = this.selectedPositions;
+        if (selectedPositions.length < positionCount) {
+            selectedPositions = new int[positionCount];
+            this.selectedPositions = selectedPositions;
+        }
+
+        // add all positions that pass the tests
+        int selectedPositionsIndex = 0;
+        for (int i = 0; i < positionCount; i++) {
+            int position = i;
+            if (testMaskBlock(maskBlock, maskBlockMayHaveNull, position) && nonNullArguments.stream().allMatch(arg -> arg.isNotNull(position))) {
+                selectedPositions[selectedPositionsIndex] = position;
+                selectedPositionsIndex++;
+            }
+        }
+        return AggregationMask.createSelectedPositions(positionCount, selectedPositions, selectedPositionsIndex);
+    }
+
+    private static boolean testMaskBlock(Block block, boolean mayHaveNulls, int position)
+    {
+        if (block == null) {
+            return true;
+        }
+        if (mayHaveNulls && block.isNull(position)) {
+            return false;
+        }
+        return block.getByte(position, 0) != 0;
+    }
+
+    private static final class NonNullArgument
```

Comment:
This name is a little confusing. If it's supposed to represent a non null argument, then why does it have a `mayHaveNull` and a `isAlwaysNull`?